### ISSUE TITLE
Bug: [no-unnecessary-condition] Even with `noUncheckedIndexedAccess`, this rule incorrectly flags `any` access on a record

### ISSUE THREAD

**Issue by tylerlaprade** [typescript-eslint/typescript-eslint#11571]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=MYewdgzgLgBAtuKALANgTwMJIIZgOYCmEAQmgArYBO2cBUBlAXDAEoGiUAmAPAFAwxolAJb4ANPxhgArnABGDSQD4YAXhgBvAL4BuXqEiwAZiBDNcaNTADkJkNb0IwydFlyES5KjToMA2nYAujAA-CHqfoF6QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false

### Repro Code

```TypeScript
const monthlyChangesByParameter: Record<
  string,
  number
  > = {};
const foo: any = 'foo';
monthlyChangesByParameter[foo] ??= [];
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUncheckedIndexedAccess": true
  }
}
```

### Expected Result

This rule should not trigger, as the Record might not (and in fact doesn't) contain this key. Especially with `noUncheckedIndexedAccess` enabled, there's no valid way to resolve this without disabling the rule.

### Actual Result

The rule has a false positive.

**Comment by Josh-Cena** [comment]
Out of curiosity what's the real world case where you have to use `any` for the key as opposed to `string`?

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/6632
It's specifically the `??=` that's flagging

**Comment by tylerlaprade** [comment]
> It's specifically the `??=` that's flagging 

FYI @bradzacher, I also ran into it with `||=`

**Comment by Josh-Cena** [comment]
@bradzacher I don't think it's a duplicate? If you remove `: any` or change it to `: string`, the error disappears. #6632 is irrelevant with `noUncheckedIndexedAccess` enabled.

**Comment by tylerlaprade** [comment]
> Out of curiosity what's the real world case where you have to use `any` for the key as opposed to `string`?

@Josh-Cena, it comes from the `data` field of a `CellValueChangedEvent` in `AgGrid`
